src/preact/component/3p-frame.js

```diff
@@ -0,0 +1,139 @@
+import {
+  getOptionalSandboxFlags,
+  getRequiredSandboxFlags,
+} from '../../core/3p-frame';
+import {parseUrlDeprecated} from '../../url';
+import {sequentialIdGenerator} from '../../utils/id-generator';
+import {useLayoutEffect, useRef, useState} from '../../../src/preact';
+
+/** @type {!Object<string,function>} 3p frames for that type. */
+export const countGenerators = {};
+
+/** @enum {string} */
+export const MessageType = {
+  // TODO(wg-bento): Add more types as they become needed.
+  EMBED_SIZE: 'embed-size',
+};
+
+// Block synchronous XHR in ad. These are very rare, but super bad for UX
+// as they block the UI thread for the arbitrary amount of time until the
+// request completes.
+const BLOCK_SYNC_XHR = "sync-xhr 'none';";
+
+// TODO(wg-bento): UA check for required flags without iframe element
+const DEFAULT_SANDBOX =
+  getRequiredSandboxFlags().join(' ') +
+  ' ' +
+  getOptionalSandboxFlags().join(' ');
+
+/**
+ * Creates the iframe for the embed. Applies correct size and passes the embed
+ * attributes to the frame via JSON inside the fragment.
+ * @param {!IframeEmbedDef.Props} props
+ * @param {{current: (!IframeEmbedDef.Api|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function ProxyIframeEmbedWithRef(
+  {
+    name: nameProp,
+    options = {},
+    messageHandler,
+    sandbox = DEFAULT_SANDBOX,
+    src: srcProp,
+    type,
+    title = type,
+    ...rest
+  },
+  ref
+) {
+  const contentRef = useRef(null);
+  if (!countGenerators[type]) {
+    countGenerators[type] = sequentialIdGenerator();
+  }
+  const [count] = useState(countGenerators[type]);
+  const [name, setName] = useState(nameProp);
+  const src = useRef(null);
+  const win = contentRef.current?.ownerDocument?.defaultView;
+  src.current = srcProp ?? (win ? getDefaultBootstrapBaseUrl(win) : null);
+
+  useLayoutEffect(() => {
+    const win = contentRef.current?.ownerDocument?.defaultView;
+    if (!win) {
+      return;
+    }
+    if (nameProp) {
+      setName(nameProp);
+      return;
+    }
+    const attrs = dict({
+      'title': title,
+      'type': type,
+      '_context': dict({
+        'location': {
+          'href': win.location.href,
+        },
+        'sentinel': generateSentinel(win),
+      }),
+    });
+    for (const key in options) {
+      attrs[key] = options[key];
+    }
+    setName(
+      JSON.stringify(
+        dict({
+          'host': parseUrlDeprecated(src.current).hostname,
+          'bootstrap': getBootstrapUrl(),
+          'type': type,
+          // "name" must be unique across iframes, so we add a count.
+          // See: https://github.com/ampproject/amphtml/pull/2955
+          'count': count,
+          'attributes': attrs,
+        })
+      )
+    );
+  }, [count, nameProp, options, title, type]);
+
+  return (
+    <IframeEmbed
+      allow={BLOCK_SYNC_XHR}
+      contentRef={contentRef}
+      messageHandler={messageHandler}
+      name={name}
+      ref={ref}
+      ready={!!name}
```

Comment:
Ok. I see the point of the `ready` now. And I think this is good. Another idea in case you think it helps: we could use `loading = UNLOAD` for this since it's already used by the `IframeEmbed` for a similar purpose.

Reply:
I tried replacing `ready={!!name}` with `loading={name ? Loading.UNLOAD : Loading.AUTO}` but found that the errors prevented by the former resurface with the change:
```
Uncaught TypeError: Cannot read property 'href' of undefined
    at frame-metadata.js:112
    at function.js:38
    at getLocation (frame-metadata.js:121)
    at draw3p (integration-lib.js:167)
    at frame.max.html:15
```

The name has to be defined as soon as the iframe src is requested because of this `once`, so used `ready` for this, not sure if there is a better/alternative way to use `loading` for the same result? : https://github.com/ampproject/amphtml/blob/7c65a3225a5c4e3b23e0a51776972685e0668dde/3p/frame-metadata.js#L59-L72